include/gsl/util

```diff
@@ -121,8 +123,8 @@ GSL_SUPPRESS(bounds.2) // NO-FORMAT: attribute
     // clang-format on
     constexpr auto at(Cont& cont, const index i) -> decltype(cont[cont.size()])
 {
-    Expects(i >= 0 && i < narrow_cast<index>(cont.size()));
     using size_type = decltype(cont.size());
+    Expects(i >= 0 && narrow_cast<size_type>(i) < cont.size());
```

Comment:
the way I interpret the assertion `static_assert(N <= std::numeric_limits<std::size_t>::max() / 2, "We only support arrays up to half the bytes of the address space.");` is that because `index` is defined as 
```cpp
// index type for all container indexes/subscripts/sizes
using index = std::ptrdiff_t;
```
then:
  a) we should not be supporting container access beyond this size. This is enforced by making the type of `i` be `index`.
  b) we should not be supporting containers that are larger than `std::ptrdiff_t` in size.
  
As per the discussion in https://github.com/microsoft/GSL/issues/1075, adding the `static_assert` for the array case is a helpful way of validating condition (b) at compile-time (in fact, I personally am more in favor of `static_assert(N <= static_cast<size_t>(std::numeric_limits<std::ptrdiff_t>::max())`). In the more general cases such as for `Cont`, I think it would make more sense to keep the `narrow_cast<index>(cont.size())` as a way of ensuring (b) at runtime. 

Imo reverting this change at the array case as well would help the code be more uniform, although not necessary due to the static_assert.